lasagne/layers/autoenc.py

```diff
@@ -0,0 +1,262 @@
+        :class:`InverseLayer` is inverse to.
+
+    Examples
+    --------
+    >>> import lasagne
+    >>> from lasagne.layers import InputLayer, Conv2DLayer, DenseLayer
+    >>> from lasagne.layers import InverseLayer
+    >>> l_in = InputLayer((100, 3, 28, 28))
+    >>> l1 = Conv2DLayer(l_in, num_filters=16, filter_size=5)
+    >>> l2 = DenseLayer(l1, num_units=20)
+    >>> l_u = InverseLayer(l2, l1)  # As Deconv2DLayer
+    """
+    def __init__(self, incoming, layer, **kwargs):
+
+        super(InverseLayer, self).__init__(
+            [incoming, layer, layer.input_layer], **kwargs)
+
+    def get_output_shape_for(self, input_shapes):
+        return input_shapes[2]
+
+    def get_output_for(self, inputs, **kwargs):
+        input, layer_out, layer_in = inputs
+        return theano.grad(None, wrt=layer_in, known_grads={layer_out: input})
+
+
+def build_autoencoder(layer, nonlinearity='same', b='new'):
+    """
+    Given a :class:`Layer` instance, this function builds a
+    symmetric autoencoder with tied weights.
+
+    Parameters
+    ----------
+    layer: a :class:`Layer` instance
+        The :class:`Layer` instance with respect to which a symmetric
+        autoencoder is built.
+    nonlinearity : 'same', list, callable, or None
+        The nonlinearity that is applied to the layer activations of theano
+        :class:`InverseLayer`. If None is provided, the layer will be linear.
+        If 'same', the :class:`InverseLayer` has the same nonlinearity as the
+        mirror :class:`Layer`.
+    b : 'new', list, Theano shared variable, numpy array, callable or None
+        An initializer for the biases. If 'new', new biases are initialized
+        using the default initialization. If a shared variable or a numpy array
+        is provided, the shape must match the incoming shape, skipping those
+        axes the biases are shared over (see below for an example). If set to
+        ``None``, the layer will have no biases and pass through its input
+        unchanged.
+        See :func:`lasagne.utils.create_param` for more information.
+
+    Returns
+    -------
+    layer: :class:`Layer` instance
+       The output :class:`Layer` of the symmetric autoencoder with
+       tied weights.
+
+    Examples
+    --------
+    >>> import lasagne
+    >>> from lasagne.layers import InputLayer, DenseLayer
+    >>> from lasagne.layers import build_autoencoder
+    >>> l_in = InputLayer((100, 20))
+    >>> l1 = DenseLayer(l_in, num_units=50)
+    >>> l2 = DenseLayer(l1, num_units=10)
+    >>> l_ae = build_autoencoder(l2, nonlinearity='same', b=None)
+    """
+
+    if isinstance(nonlinearity, (tuple, list)):
+        n_idx = 0
+
+    if isinstance(b, (tuple, list)):
+        b_idx = 0
+
+    layer = unfold_bias_and_nonlinearity_layers(layer)
+    layers = get_all_layers(layer)
+    autoencoder_layers = [layer]
+
+    for layer in layers[::-1]:
+        if not isinstance(layer, InputLayer):
```

Comment:
Instead of indenting everything underneath, it would be easier to just have a:
```
if isinstance(layer, InputLayer):
    continue
```
in the beginning! That makes it easier to understand (I had to scroll a long way down to check if there is an `else` case for this `if` statement).